Add tutorial about initrd structure and build process in mkosi (#4067)
systemd/mkosi: docs/initrd_generation.md

apply these updates very early in the boot process.

The remainder of the initrd contains the early userspace, loadable kernel
modules, and firmware files. In early Linux systems, these files were stored
as a filesystem image, typically using ext2. This approach was later replaced
by the use of an initramfs, which stores files in one or more cpio archives
rather than in a filesystem image. Despite the terminology, what is commonly
called an “initrd” today usually contains an initramfs rather than an ext2
initrd image.

From the beginning, the
[kernel code](https://github.com/torvalds/Linux/blob/1da177e4c3f41524e886b7f1b8a0c1fc7321cac2/init/initramfs.c#L511)
responsible for extracting the initramfs has
[supported](https://www.kernel.org/doc/html/v6.18/driver-api/early-userspace/buffer-format.html)
multiple concatenated cpio archives, each optionally compressed. These
archives are unpacked sequentially to populate the in-memory `rootfs` used
during early boot.

`mkosi` uses this mechanism to split the initramfs into distinct parts:

* **Default initrd**: contains only the early userspace, primarily `init` and its
  supporting files. This archive is kernel-version independent.

> * **Kernel Module initrd (KMI)**: contains kernel modules and firmware files, which

Reviewer: The current usage in the codebase is plural
```suggestion
* **Kernel Modules initrd (KMI)**: contains kernel modules and firmware files, which
```

  are specific to a particular kernel version.

For historical reasons, commonly used initrd inspection tools such as
`lsinitrd`, `unmkinitramfs`, and even newer tools like `3cpio`, only process the
early cpio and the first cpio archive in the initrd. Any additional concatenated
archives are ignored.

As a result, kernel modules included in the KMI do not appear when inspecting a
typical `mkosi`-generated initrd with these tools.

> # Inspecting the Kernel Module initrd

Reviewer: ```suggestion
# Inspecting the Kernel Modules Initrd
```

To work around the limitations of existing tools, `mkosi` provides the
`SplitArtifacts=kernel-modules-initrd` option. When enabled, `mkosi` saves the
KMI as a separate file, making it directly inspectable.

The KMI is a cpio archive, optionally compressed, and can be examined using
standard tools:

```
# Rebuild and generate a separate KMI file

> $ mkosi --split-artifacts=kernel-modules-initrd -f

Reviewer: This is not really helpful for somebody trying to follow along. In an empty directory it doesn't produce the split output and in the mkosi repo it will be `mkosi.output/image.kernel-modules-initrd`.

# List KMI contents using lsinitrd
$ lsinitrd my_image.kernel-modules-initrd

# Or inspect directly using cpio
$ cat my_image.kernel-modules-initrd | cpio -itv
```

# How mkosi Builds the initrd

There are some nuances to the way `mkosi` generates the KMI which you should
familiarize yourself with.

The "default initrd" is built as a standalone image. In contrast, the KMI is
generated as a sub-step of the main image build. During this step, `mkosi`
examines the filesystem of the main image to locate kernel modules and firmware
files to include in the KMI. Consequently, if required firmware packages are not

> listed in the main configuration’s `Packages=` option, they will not be present

Reviewer: ```suggestion
listed in the main configuration's `Packages=` option, they will not be present
```

in the KMI and will be unavailable during early boot.

Kernel module selection is controlled by two options:

* `KernelModules=` determines which modules are included in the main image.
* `InitrdKernelModules=` determines which modules are included in the KMI.

Both options support glob patterns to specify inclusion and exclusion rules.
Module filtering is performed relative to the kernel modules present in the

> main image filesystem. Module and firmware Dependencies of selected modules

Reviewer: might be a directory
```suggestion
main image. Module and firmware dependencies of selected modules
```

Author: This refers to build time, not the output format. Corrected case.

are resolved automatically and included as needed.

To reduce image size, `mkosi` removes unused kernel modules and firmware files.
Because `KernelModules=` is processed first, any modules not selected are deleted
after dependency resolution. When the KMI build step later processes
`InitrdKernelModules=`, it can only select from modules that survived this earlier
pruning. As a result, the KMI is limited to modules already included in
`KernelModules=`. This behavior is a known limitation and may be addressed in
future versions.

The "default initrd" configuration shipped with `mkosi` specifies a default set of
modules via `KernelModules=`. Although this configuration appears to belong to
the "default initrd", it actually determines which modules are included in the KMI
Kernel modules are never included in the "default initrd" itself, as it is intended
to remain kernel-independent.

Firmware inclusion is controlled using the `FirmwareFiles=` option. At present,

> this option applies uniformly to both the main image and the KMI; You cannot

Reviewer: ```suggestion
this option applies uniformly to both the main image and the KMI. You cannot
```
or
```suggestion
this option applies uniformly to both the main image and the KMI; you cannot
```

component, commonly referred to as the *early cpio*, is an optional, uncompressed
cpio archive containing CPU microcode updates. The kernel attempts to locate and
apply these updates very early in the boot process.

The remainder of the initrd contains the early userspace, loadable kernel
modules, and firmware files. In early Linux systems, these files were stored
as a filesystem image, typically using ext2. This approach was later replaced
by the use of an initramfs, which stores files in one or more cpio archives
rather than in a filesystem image. Despite the terminology, what is commonly
called an “initrd” today usually contains an initramfs rather than an ext2
initrd image.

From the beginning, the
[kernel code](https://github.com/torvalds/Linux/blob/1da177e4c3f41524e886b7f1b8a0c1fc7321cac2/init/initramfs.c#L511)
responsible for extracting the initramfs has
[supported](https://www.kernel.org/doc/html/v6.18/driver-api/early-userspace/buffer-format.html)
multiple concatenated cpio archives, each optionally compressed. These
archives are unpacked sequentially to populate the in-memory `rootfs` used
during early boot.

> `mkosi` uses on this mechanism to separate the initramfs into separate cpio archives

Reviewer 2: "on this mechanism" doesn't sound right
